 ### חלק א׳ - חקירת מערך נתונים

In [1]:
#Import libraries:
import pandas as pd  # pa
import numpy as np  # numpy is a package for working with numerical data

In [2]:
#Reading the 'HW1_data.csv' file:
df_soccer = pd.read_csv('HW1_data.csv')

## Question 1:

In [3]:
#1

df_soccer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 684 entries, 0 to 683
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   league    684 non-null    object 
 1   year      684 non-null    int64  
 2   position  684 non-null    int64  
 3   team      684 non-null    object 
 4   matches   684 non-null    int64  
 5   wins      684 non-null    int64  
 6   draws     684 non-null    int64  
 7   loses     684 non-null    int64  
 8   scored    684 non-null    int64  
 9   conceded  681 non-null    float64
 10  pts       684 non-null    int64  
dtypes: float64(1), int64(8), object(2)
memory usage: 58.9+ KB


We can see according to the "info()" of the data frame that there are 684 entries.

## Question 2:

In [4]:
#2

df_soccer.info()
display(df_soccer)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 684 entries, 0 to 683
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   league    684 non-null    object 
 1   year      684 non-null    int64  
 2   position  684 non-null    int64  
 3   team      684 non-null    object 
 4   matches   684 non-null    int64  
 5   wins      684 non-null    int64  
 6   draws     684 non-null    int64  
 7   loses     684 non-null    int64  
 8   scored    684 non-null    int64  
 9   conceded  681 non-null    float64
 10  pts       684 non-null    int64  
dtypes: float64(1), int64(8), object(2)
memory usage: 58.9+ KB


league  year  position                   team  matches  wins  draws  \
0    La_liga  2014         1              Barcelona       38    30      4   
1    La_liga  2014         2            Real Madrid       38    30      2   
2    La_liga  2014         3        Atletico Madrid       38    23      9   
3    La_liga  2014         4               Valencia       38    22     11   
4    La_liga  2014         5                Sevilla       38    23      7   
..       ...   ...       ...                    ...      ...   ...    ...   
679     RFPL  2019        12              PFC Sochi       30     8      9   
680     RFPL  2019        13              FK Akhmat       30     7     10   
681     RFPL  2019        14  Krylya Sovetov Samara       30     8      7   
682     RFPL  2019        15              FC Tambov       30     9      4   
683     RFPL  2019        16            FC Orenburg       30     7      6   

     loses  scored  conceded  pts  
0        4     110      21.0   94  
1        6     118      38.0   92  
2        6      67      29.0   78  
3        5      70      32.0   77  
4        8      71      45.0   76  
..     ...     ...       ...  ...  
679     13      40      39.0   33  
680     13      27      46.0   31  
681     15      33      40.0   31  
682     17      37      41.0   31  
683     17      28      52.0   27  

[684 rows x 11 columns]

We can see according to the "info()" of the data frame that there are two fields of type "object", so we can be asure that they are categorical.
About the other columns, we can see according to the data frame table that the other fields are continuous, so thay are numerical.
To sum up:

The fields that are categorical are:
1. league
2. team

The fields that are numerical are:
1. year
2. position
3. matches
4. wins
5. draws
6. loses
7. scored
8. conceded
9. pts

## Question 3:

In [5]:
#3

print(len(df_soccer.league.unique()))
print()
print(len(df_soccer.team.unique()))

6

168


Number of unique values at "league" field: 6

Number of unique values at "team" field: 168

## Question 4:

In [6]:
#4
print("Let's find out if there are missing values at our data frame: " + str(pd.isna(None)) + ".") 
for i in df_soccer.columns:
    if df_soccer.isna().sum()[i] != 0:
        print('The field', i, 'have', df_soccer.isna().sum()[i], 'missing values.')

Let's find out if there are missing values at our data frame: True.
The field conceded have 3 missing values.


The field that has a null vlaues is conceded, and it has 3 missing values.

## Question 5:

In [7]:
#5

min_league = ''
temp = 0
flag = True

#Loop over array of the leagues, find the number of different teams for each league and find out witch league has the fewer teams.  
for diff_league in df_soccer.league.unique():
    if flag:
       temp = len(df_soccer[df_soccer.league == diff_league].team.unique())
       flag = False
    print(diff_league + ':',len(df_soccer[df_soccer.league == diff_league].team.unique()))
    if len(df_soccer[df_soccer.league == diff_league].team.unique()) < temp:
     temp = len(df_soccer[df_soccer.league == diff_league].team.unique())
     min_league = diff_league
print()
print("min:", min_league)

La_liga: 30
EPL: 30
Bundesliga: 24
Serie_A: 30
Ligue_1: 29
RFPL: 25

min: Bundesliga


Number of different teams over the years of:
La_liga: 30
EPL: 30
Bundesliga: 24
Serie_A: 30
Ligue_1: 29
RFPL: 25

The league that has the least number of teams over the years is: Bundesliga

## Question 6:

In [ ]:
#6

#Method for calculate the avarge of scored goals per year.
def calculateAvgGoals(row):
    avg = row.scored/row.matches
    return avg

#Adds new columne to the data frame that store the avarge scored goals of each team per year.
df_soccer["avg_goals"] = df_soccer.apply(calculateAvgGoals, axis = 'columns')

#Creates a groupby object of 'team'.
gb_team = df_soccer.groupby('team')

#Prints the columns that describe the avarge scored goals of each team for every year.
for team, team_df_soccer in gb_team:
    print(team)
    display(team_df_soccer[['year', 'avg_goals']])

In [9]:
#checking the correlation 
print(df_soccer.position.corr(df_soccer.avg_goals))

-0.7587818572237601


The corilation we have got(-0.7587) tells us that there is a liniar corilation(-1<-0.7587<0),
and even a strong corilation according to the high value of the corilation.

## Question 7:

In [10]:
#7

#Creates a groupby object of 'league' and 'year', and calculate the median value of 'scored' for each league and year.
gb_median = df_soccer.groupby(['league', 'year']).median()['scored']

#Prints the groupby object we have created.
print(gb_median)

league      year
Bundesliga  2014    44.0
            2015    44.0
            2016    44.5
            2017    43.5
            2018    53.0
            2019    48.0
EPL         2014    46.5
            2015    48.5
            2016    47.5
            2017    44.5
            2018    51.5
            2019    46.5
La_liga     2014    43.0
            2015    45.5
            2016    53.0
            2017    46.5
            2018    47.0
            2019    46.0
Ligue_1     2014    44.0
            2015    46.0
            2016    42.5
            2017    45.5
            2018    46.0
            2019    33.0
RFPL        2014    31.0
            2015    34.5
            2016    30.5
            2017    31.5
            2018    30.5
            2019    36.5
Serie_A     2014    49.0
            2015    46.5
            2016    56.0
            2017    51.0
            2018    51.5
            2019    51.5
Name: scored, dtype: float64


C:\Users\liorb\AppData\Local\Temp\ipykernel_24608\2117126896.py:4: FutureWarning: The default value of numeric_only in DataFrameGroupBy.median is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  gb_median = df_soccer.groupby(['league', 'year']).median()['scored']


We have found that the league that has the highest difference between the medians over the years is "ligue_1"(differnce of 13 goals scored(46 is the highest at year 2018/2015 and 33 is the lowest at year 2019).

## Question 8:

In [11]:
#8

#Method for calculate the sum of pts each team got at each year.
def calculatePts(row):
    sumOfPts = row.wins*3 + row.draws*1
    return sumOfPts

#Adds new columne to the data frame that store the sum of pts each team got at each year.
df_soccer['our_pts'] = df_soccer.apply(calculatePts, axis='columns')

#Method for compare our values of pts to the values that given to us.
def comparePts(row):
    if row.pts == row.our_pts:
        return True
    else:
        return None
    
#Adds new columne to the data frame that store boolian value - "True" if our values of pts to the values that given to us is equal, else - "None".
df_soccer['compare_pts'] = df_soccer.apply(comparePts, axis='columns')
display(df_soccer[['year', 'team', 'matches', 'wins', 'draws', 'loses', 'pts', 'our_pts', 'compare_pts']])


df_soccer.info()
print()
#we can use the `count()` method to count the number of non-NA values of 'compare_pts' column.
print("the number of non-NA values of 'compare_pts' column is: " + str(df_soccer.count()['compare_pts']) + " of total 684 entries.")
print()
print("We will Find information about the mistake:")
display(df_soccer[df_soccer['compare_pts'].isna() == True])

print('the mistake happens to be in league is "Serie_A", year 2017, the team "Crotone".')

year                   team  matches  wins  draws  loses  pts  our_pts  \
0    2014              Barcelona       38    30      4      4   94       94   
1    2014            Real Madrid       38    30      2      6   92       92   
2    2014        Atletico Madrid       38    23      9      6   78       78   
3    2014               Valencia       38    22     11      5   77       77   
4    2014                Sevilla       38    23      7      8   76       76   
..    ...                    ...      ...   ...    ...    ...  ...      ...   
679  2019              PFC Sochi       30     8      9     13   33       33   
680  2019              FK Akhmat       30     7     10     13   31       31   
681  2019  Krylya Sovetov Samara       30     8      7     15   31       31   
682  2019              FC Tambov       30     9      4     17   31       31   
683  2019            FC Orenburg       30     7      6     17   27       27   

    compare_pts  
0          True  
1          True  
2          True  
3          True  
4          True  
..          ...  
679        True  
680        True  
681        True  
682        True  
683        True  

[684 rows x 9 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 684 entries, 0 to 683
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   league       684 non-null    object 
 1   year         684 non-null    int64  
 2   position     684 non-null    int64  
 3   team         684 non-null    object 
 4   matches      684 non-null    int64  
 5   wins         684 non-null    int64  
 6   draws        684 non-null    int64  
 7   loses        684 non-null    int64  
 8   scored       684 non-null    int64  
 9   conceded     681 non-null    float64
 10  pts          684 non-null    int64  
 11  avg_goals    684 non-null    float64
 12  our_pts      684 non-null    int64  
 13  compare_pts  683 non-null    object 
dtypes: float64(2), int64(9), object(3)
memory usage: 74.9+ KB

the number of non-NA values of 'compare_pts' column is: 683 of total 684 entries.

We will Find information about the mistake:


league  year  position     team  matches  wins  draws  loses  scored  \
425  Serie_A  2017        18  Crotone       38     9      8     21      40   

     conceded  pts  avg_goals  our_pts compare_pts  
425      66.0  180   1.052632       35        None

the mistake happens to be in league is "Serie_A", year 2017, the team "Crotone".


According to that the "Serie_A" league have 38 matches per year according to the information we got, and the max number that a team at that league
and yaer can got is 38*3=116(if the team won all the games at that year), the given value of "pts" we have got is not possible,
so the origin of the mistake is the given "pts" column

## Question 9:

In [12]:
#9

#Method for compare the values of conceded goals to the values of scored goals.
def isFailure(row):
    if row.conceded != None:
            if row.conceded > row.scored:
                return True
            else:
                return None
    
#Adds new columne to the data frame that store boolian value - "True" if a team is conceded more goals than scored, else - "None".
df_soccer['is_failure'] = df_soccer.apply(isFailure, axis='columns')

#Gets an array that store the different leagues in the data frame.
arrOfLeagues = df_soccer.league.unique()

listCountFailure = []
#find the number of "failed teams" for over the years of each league and append to the empty list we have created.
for i in arrOfLeagues:
    filt = (df_soccer['league'] == i) & (df_soccer['is_failure'] == True)
    val = df_soccer.loc[filt].count()['is_failure']
    leagueAndVal = i + " " + str(val)
    listCountFailure.append(leagueAndVal)
    
display(listCountFailure)

['La_liga 75',
 'EPL 71',
 'Bundesliga 63',
 'Serie_A 63',
 'Ligue_1 69',
 'RFPL 52']

According to the results we got, the league with the most "failed teams" for over the years is "La_liga" league. 